### PR TITLE
feat(server): per-account, per-region Kura egress floor and ceiling in /ops

### PR DESCRIPTION
**Pull request by esnunes**
## What

Staff can give an account its own Kura egress floor and ceiling in each region, from the account's `/ops` page, the way they already override its disk claim. One row per region the account holds an instance in, showing what applies now, what it falls back to, and how far the box it sits on will let the floor go.

## Why

A tenant's shaping came only from its region's constants. An account whose restores were starving a shared box, or one being given room for a migration, could only be retuned by changing what every account in that region gets.

It is scoped **per region** because the boxes are. The budget is declared per machine, so a floor a 3 Gbit/s box can keep is one a 1 Gbit/s box cannot, and a single account-wide number would either waste the roomy regions or promise what the small ones cannot deliver.

## The three numbers, and which one is a bound

- **The operator's override** is the intent. Whichever half staff state is the number that applies.
- **The region's pair** is a default sized for the fleet, not a statement about the hardware. It fills whichever half is left blank and gives way where the two disagree — a stated floor of 2000 against a defaulted ceiling of 1500 means 2000 of headroom was wanted, not that 1500 was mistyped. A pair the operator states *both* halves of is theirs to keep coherent, so an inverted one is rejected rather than silently resolved.
- **The box** is the only real bound, and it is measured rather than assumed.

## How it reaches the pod

Through the two fields that already carry the region's numbers, not through new ones:

| number | KuraInstance field | cascades to |
|---|---|---|
| floor | `egressGuaranteedMbps` | pod `tuist.dev/egress-mbps` request/limit **and** `tuist.dev/egress-class` `floor_mbps` |
| ceiling | `podAnnotations["kubernetes.io/egress-bandwidth"]` | Cilium's pacing **and** `tuist.dev/egress-class` `burst_mbps` |

The server renders both at the account's effective rates; everything downstream is unchanged, and the kura-controller is untouched by this PR. One number per knob means the slice the scheduler reserves, the ceiling Cilium paces, and the class the shaper builds can never describe different limits.

A change reaches the account's instances **in that region only** — the override row is keyed `(account, region)`, and the reconciler re-renders that region's instance.

Both fields are pod-spec state, so applying a change recreates that region's replicas. They keep their volumes (a `data-<sts>-<ordinal>` claim belongs to the ordinal, not the pod), so a replica reopens the same warm cache and the cost is a restart behind the standby, not a refill. The form says so, and asks before saving.

## What bounds a floor

Not what the region's boxes advertise — that only catches a number larger than the NIC. A floor is the pod's `tuist.dev/egress-mbps` request, so **every replica reserves it** and the box has to hold all of them at once. And the box is fixed: the replica's volume pins it there, so there is no other node to try.

The rollout gets there one replica at a time, each step handing in the old reservation before asking for the new. That makes the last step binding and cancels out what the replicas hold on the way:

```
step k:  O + k·new + Σ_{i>k} c_i  ≤  allocatable
step r:  O + r·new                ≤  allocatable

         replicas × floor  ≤  allocatable − other tenants
```

Deliberately *not* written in terms of what the replicas hold now: an account mid-rollout holds the new floor on one replica and the old one on the other, so a rule stated in those terms answers differently depending on when it is asked.

Reductions are never checked — the final state reserves less than the box already carries, and every step frees more than it takes. Nor is the ceiling, which reserves nothing.

## Measuring the box

One box is read — the one the account's instance sits on, found from its own pods — in three calls, whatever the region's size:

```
list_pods(kura, managed-by=…,region=<r>,account=<a>)   which box, its replicas, its own reservation
get_node(<that box>)                                   allocatable
list_pods_on_node(<that box>)                          everything reserved on it
```

The last is cluster-wide on purpose: an extended resource is reserved against the node, not a namespace, so anything on the box counts whoever owns it. Scoping the first to the account's instance in this region is the label selector's job — a box can carry the same account's pods from another region or from a self-hosted deployment, and those are neither its replicas nor its reservation.

There is no API for what a node has allocated. `kubectl describe node` builds its "Allocated resources" table from the same request (`fieldSelector=spec.nodeName=<node>`, terminal phases excluded) and sums the requests itself, because allocation is a property of the pods bound to a node while `node.status` is written by the kubelet, which is not the component that places them.

The replica count is raised to what the region declares, so a replica deleted and not yet recreated is still counted — it is in no pod list, but its volume brings it back to this box.

A box that cannot be read yields `nil`, which falls back to the advertised budget: reporting it as roomier than it is would be worse than reporting it as unknown. That answer is cached like any other — Cachex stores a bare `nil` and reads it back as a miss, so the measurement is wrapped in a tuple to keep an unreadable box from being re-measured on every render. The reads are unlocked (Cachex serialises `locking: true` through one process, so a slow apiserver would park every other cache read behind a page render) and bounded by an explicit 5s timeout with retries off.

Needs cluster-wide `pods: [list]`, which the server gets through a ClusterRole of its own (`<server>-cluster-reads`: `nodes get/list` + `pods list`) rather than by joining the controller's `-node-reads` role, which also carries `persistentvolumes update/patch`. The server is the internet-facing component; read-only is all it needs. That list does expose every namespace's pod specs — inline env values and args — to it, which is a conscious grant.

The check is against a live reading, not a guarantee — another account can be provisioned onto the box between saving and rolling. It catches the mistake where it is made; the 30-minute `kube_pod_status_unschedulable` alert stays the backstop for the race.

## Entitlement

The region's own floor stays Enterprise-only, and the default column and placeholder show that rather than the region's raw number — an Air account reserves nothing, and a form promising it 25 Mbps would describe an instance that does not exist. An override outranks the gate: it is a staff decision about one account made with the box in front of them, while the entitlement is what happens when nobody has made one.

## The manifest revision keys on the rendered pair

The reconciler converges on `(image_tag, manifest_revision)` alone, so the revision is a function of what the manifest renders rather than of the override alone. The override is one of three inputs — the region's own numbers and the entitlement gating the floor are the others — and any of them moving has to move the revision or the change sits unapplied until something unrelated does.

Re-applying is cheap where the pair has not moved: the rendered spec is identical, so the CR is written once with a new annotation and no pod is touched. On rollout every governed instance gains a `+egress<pair>` suffix and re-applies once, with no pod churn; the cloud regions render no suffix and do not move at all.

One deliberate consequence: a plan change across the Enterprise boundary now rolls an account's cache pods, because the floor is a pod request. That is what makes the entitlement real rather than decorative.

## The form

- The **Node limit** column reads `500 Mbps` / `up to 175` — the box, and how far this account's floor can go on it. The floor input carries that as its `max`, so the browser refuses the number in its own chrome before the row is submitted.
- The result of a Save renders **inside the card**, under the table it is about, as one message for the whole Save: it names every region changed and totals the instances. A rejection names the regions too — with several rows on screen, "could not be updated" leaves an operator hunting for which one the form refused. A Save that changes nothing says so, rather than reporting a rollout that is not happening.
- Only rows whose values moved are validated and written. The reading behind the check can shift under a table an operator is looking at, and a row they did not touch must not refuse the save of the one they did. A failing write is reported rather than folded into "nothing changed", and an instance that holds no pods gets no row at all.
- Saving asks first, via `data-confirm` on the **button**. Not on the form: LiveView 1.2 has no `data-confirm` handling of its own, and phoenix_html's polyfill walks up the DOM from whatever was clicked and confirms on the first ancestor carrying the attribute — on a form that is every click into a field.

## Also in here

`infra/egress-tree-agent` clamped a tenant's ceiling to the node budget but took its floor as given. HTB does no admission control, and `classRates` raises the ceiling to meet a higher floor, so a floor above the whole box carried the ceiling straight past the cap the tree exists to hold: a 2000 Mbps floor on a 1000 Mbps box yields `rate 2Gbit ceil 2Gbit`. The floor is now clamped first.

## Validation

7374 of 7378 server tests pass locally; the four failures are pre-existing locale tests (they need non-English locales compiled, which the test env skips by default). The tests added here cover the bound and its stability mid-rollout, `Capacity` for node-wide reserved / the account-and-region selector / a replica mid-recreate / an unreadable box, the provisioner for what the manifest renders and when the revision moves, and LiveView for the per-region table, the in-card result across two regions, and the confirm. `mix credo` is clean; the kura-controller builds and its tests pass.

End to end on the k01 cluster, both regions, both halves:

```
us-east    (nunes-desktop, 1 Gbit root)   1:2d03 rate 60Mbit ceil 400Mbit
eu-central (microvm01,    500 Mbit root)  1:2d03 rate 25Mbit ceil 100Mbit
```

Saved through the same call the form makes; the CR moved within seconds of the Oban nudge, the StatefulSet template picked up `tuist.dev/egress-mbps: 60`, `400M`, and `{"classid":"1:2d03","floor_mbps":60,"burst_mbps":400}`, the pods rolled onto it, and the HTB class on the box matched. eu-central exercised the half-override path (ceiling only, floor left to the region).

Refusals checked live against real node budgets:

```
floor 2000 on a 1000 Mbps box: must not exceed the box's 1000 Mbps
ceiling 900 on a 500 Mbps box: must not exceed the box's 500 Mbps
inverted 900/100:              must not exceed the ceiling
```

The headroom reading, derived from the cluster with nothing hardcoded, on a box shared with a second tenant holding 150 Mbps:

```
eu-central  microvm01      allocatable 500   reserved 200  own 50   -> available 350,  2 replicas -> up to 175
us-east     nunes-desktop  allocatable 1000  reserved 400  own 400  -> available 1000, 2 replicas -> up to 500
```

and the form refusing on those numbers: `176 -> "must be at most 175 Mbps on microvm01"`, with 175 accepted. Saving 120/300 there put `egressGuaranteedMbps: 120` and `300M` on the CR, both replicas onto `request 120` / `300M` / `{"floor_mbps":120,"burst_mbps":300}`, and the box's class at `rate 120Mbit ceil 300Mbit` under a 500 Mbit root.

RBAC checked on the cluster after the split: the server SA answers `no` to `patch persistentvolumes`, `update persistentvolumes` and `watch nodes`, and `yes` to `get/list nodes` and `list pods`.

The wedge and its recovery were both exercised: a floor of 480 on the 500 Mbps box left replica-1 `Pending` while replica-0 kept serving on the old revision (`updated=1 ready=1`), and clearing the override rescheduled it onto the same node and the same PVC, with no data rebuilt.

## Known limitation

Reserved is summed from `spec.containers` only, not `initContainers`. Kubernetes' effective pod request is `max(sum of containers, max of any init container)`, so a pod whose init container asked for more egress than its main ones would be under-counted. No Kura pod does that, and an init container requesting an egress extended resource would be unusual, but it is an assumption rather than a guarantee.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: server/lib/tuist/kura/capacity.ex [hunk 1]
```diff
@@ -153,6 +156,165 @@ defmodule Tuist.Kura.Capacity do
...
+      KeyValueStore.get_or_update(key, [ttl: to_timeout(minute: 1), locking: false], fn -> {:ok, measure.()} end)
+
+    value
+  end
+
+  defp measure_egress_headroom(region_id, account_handle) do
+    with {:ok, pods} <- Client.list_pods(@namespace, account_selector(region_id, account_handle), timeout: @read_timeout),
+         [pod | _] <- Enum.filter(pods, &(pod_node_name(&1) && not terminal?(&1))),
+         node = pod_node_name(pod),
+         on_box = Enum.filter(pods, &(pod_node_name(&1) == node)),
```
Comment (pepicrft) [inline server/lib/tuist/kura/capacity.ex]: Could we restore the per-node calculation here? The controller deliberately makes affinity preferred rather than required, so an account can be split across nodes and each replica remains pinned to its node by its persistent volume claim. I reproduced that layout: with the roomy node returned first at 1,000 megabits per second available and its sibling on a constrained node at 100, this reports the roomy node and allows a floor that cannot be placed on the constrained node. The earlier implementation grouped the placed replicas and used the most constrained node. Deriving `on_box` from the placed list also keeps terminal pods out of the reservation and replica count.

**Review by pepicrft — APPROVED** [review]
Approved as requested. I validated the split-placement case with a focused regression test and left the verified edge case inline.

**Comment by fortmarek** [comment]
Two things I would want fixed before this lands, plus a forward-looking note.

### 1. `egress_headroom/2` overstates the box, and can report more available than the node has

`server/lib/tuist/kura/capacity.ex:250`

`on_box` filters only on `nodeName`, but `Client.list_pods_on_node/2` excludes `Succeeded`/`Failed` through its fieldSelector. A terminal replica of the account on that box therefore enters `own_mbps` without ever having entered `reserved`, and `allocatable - (reserved - own_mbps)` adds back a reservation nobody holds. `measure_reserved_gib/1` two functions up already rejects terminal pods for exactly this reason ("the scheduler counts neither against the node").

Measured against a 1000 Mbps box carrying one Running replica at 100, one Failed replica at 400, and a neighbour at 200:

```
WITH EVICTED: %{node: "box-1", replicas: 2, allocatable_mbps: 1000, available_mbps: 1200}
MAX FLOOR (with evicted): 600
TRUTH:        %{node: "box-1", replicas: 2, allocatable_mbps: 1000, available_mbps: 800}
MAX FLOOR (truth): 400
```

`available_mbps` above `allocatable_mbps` is the tell. The form would then accept a floor of 600 across 2 replicas, reserving 1200 on a 1000 Mbps box, which is precisely the Pending wedge the check exists to prevent. Evicted pods are the common case on a disk-pressured bare-metal box, and they linger in the API until deleted.

Fix is `Enum.filter(pods, &(pod_node_name(&1) == node and not terminal?(&1)))`, which also stops line 264 counting a terminal pod as a replica.

### 2. "Nothing was saved" is asserted after some regions were saved

`server/lib/tuist_web/live/ops_account_live.ex:511`

The pre-flight cast covers every row before any write, but the writes run one region at a time, each in its own `Repo.transaction`, with no compensation. If region B's write fails after region A's committed, the banner reads "Nothing was saved: EU Central was rejected" while US East is written and its pods are already rolling.

It is reachable because `update_egress_limits_override/3` re-casts against `Capacity` readings whose cache is only a minute long, so the pre-flight cast and the write can disagree. This is the same class of mistake the description says the message design exists to avoid ("A Save that changes nothing says so, rather than reporting a rollout that is not happening"). Either report `written` alongside `failed`, or fail the whole submit before any write. There is no test on this path today: the LiveView tests cover the success and the no-change messages only.

### Future: scale the floor from demand

Not for this PR. The floor is the number worth automating later, for the same reason [spec #87](https://hive.tuist.dev/specs/87) gives for claim sizes: the region constant is a weak predictor of any individual account, and the override table is operators doing by hand what telemetry should decide. It is also the expensive half to get wrong, being the request every replica reserves, while the ceiling reserves nothing and HTB already lets a class borrow up to it when the box is quiet.

The prerequisite is observation, which is missing today, so staff setting these overrides are tuning blind. The agent already exports `kura_egress_tree_class_{sent_bytes,drops,backlog_bytes}`, but `infra/egress-tree-agent/internal/agent/tree.go:153` parses only `bytes`, `drops` and `backlog`, dropping HTB's `lended` and `borrowed`, which are the "hit its rate and borrowed" and "lent to its neighbours" signals. The series are also labelled by `classid` alone, so nothing account-level can be built from them without joining the `kura.tuist.dev/egress-class-id` CR annotation. Both are small additions, and useful immediately for the manual overrides this PR ships.

If it ever does become automatic, spec #87 suggests it belongs as a fourth output of the same decision core, on the same proposals record and the same supervised-then-automatic ladder as placement, instance count and claim size, rather than as a loop of its own.

**Review by fortmarek — APPROVED** [review]

File: server/lib/tuist/kura/capacity.ex (hunk 1 above)
Comment (esnunes) [reply server/lib/tuist/kura/capacity.ex]: Restored, and you were right that it is reachable rather than theoretical — thanks for reproducing it.

Two changes on top of what it was before:

**The binding box is now chosen by what it can hold per replica**, not by what it has left. An account placed unevenly rebuilds a different number of replicas on each box, so `available` alone picks the wrong one. Your scenario is a test: roomy box at 1000 available, sibling constrained, and it reports `constrained`.

**`on_box` comes from the placed list**, as you suggested — that also fixes a real bug @fortmarek spotted independently: terminal pods are excluded from a box's `reserved` by the field selector, so counting one as the account's own added back a reservation nobody holds and could report more available than the box has.

Two follow-ons from the same thread:

- The reads this needs are now *rejected* on the reconcile path (`Mimic.reject` on `Capacity.egress_headroom/2`, `egress_budget_mbps/1`, `Client.get_node/2`, `list_pods_on_node/2` in two provisioner tests). Verified it fails when a read is added. A cluster that cannot be read should cost an operator a bound on a form, never an instance its limits.
- The form now says when this is happening: the Node limit column reads `up to 100, across 2 boxes` for a split account, because otherwise the lower bound has no visible cause — the replicas being in two places is exactly what the operator cannot see.

**Comment by esnunes** [comment]
Both fixed. Thanks — #1 in particular was a real defect and the repro made it obvious.

### 1. Terminal pods — fixed

Exactly as you diagnosed: `on_box` was filtered from the raw account list while `reserved` comes from a list the field selector already narrows to non-terminal pods, so a terminal replica entered the account's own reservation without ever entering the box's. It came in when I moved the phase filter into the selector and dropped the client-side one, without noticing the other list still needed it.

`on_box` now derives from the placed (non-terminal) list, which also stops a terminal pod counting as a replica. Your numbers are the test — 1000 Mbps box, Running replica at 100, Failed at 400, neighbour at 200 — asserting `available_mbps: 800, replicas: 2` where it reported 1200.

### 2. Partial saves — fixed

Right that the pre-flight cast does not prevent it, and for the reason you give: the write re-casts against a reading whose cache is a minute long, so the two can disagree.

The banner now carries both halves — *"Egress limits updated in US East. 1 instance is recreated to pick them up. EU Central was rejected and not saved."*

I kept the two rejection messages distinct, which your note did not ask for but seemed to matter: the pre-flight path still says *"Nothing was saved: EU Central was rejected"*, asserting it of the whole table, because nothing is written there. Collapsing them into one wording would have made the pre-flight case understate what was skipped. New test covers the partial path by failing one region's write; it had no coverage, as you said.

### 3. Demand-scaled floors — agreed, not here

Agreed on all of it, including that the floor is the expensive half and the one worth automating. Not in this PR.

The observability gap is the part I would want first regardless of automation, since it is what makes the manual overrides this ships less blind: `lended`/`borrowed` are the "hit its rate" and "lent to its neighbours" signals and the parser drops them, and `classid`-only labels mean nothing account-level can be built without joining the CR annotation. Both small. Happy to pick that up as a follow-up if nobody else has it.
